Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm

```diff
@@ -46,150 +47,460 @@ - (void)resetUpcomingSampleBufferPresentationTimeExpectations;
-            NSDictionary *userInfoDict = @{ (__bridge NSString *)AVSampleBufferDisplayLayerFailedToDecodeNotificationErrorKey: (__bridge NSError *)error.get() };
-            [NSNotificationCenter.defaultCenter postNotificationName:AVSampleBufferDisplayLayerFailedToDecodeNotification object:m_renderer.get() userInfo:userInfoDict];
-            [NSNotificationCenter.defaultCenter postNotificationName:AVSampleBufferVideoRendererDidFailToDecodeNotification object:m_renderer.get() userInfo:userInfoDict];
-        }
-    });
 
     resetReadyForMoreSample();
 }
 
+void VideoMediaSampleRenderer::decodedFrameAvailable(RetainPtr<CMSampleBufferRef>&& sample)
+{
+    assertIsCurrent(m_workQueue);
+
+    assignResourceOwner(sample.get());
+
+    if (m_timebase) {
+        purgeDecodedSampleQueue();
+        PAL::CMBufferQueueEnqueue(ensureDecodedSampleQueue(), sample.get());
+    }
+
+    if (m_renderer)
+        [m_renderer enqueueSampleBuffer:sample.get()];
+    else if (m_displayLayer) {
+        callOnMainThread([weakThis = ThreadSafeWeakPtr { *this }, sample = sample] {
+            ALLOW_DEPRECATED_DECLARATIONS_BEGIN
+            if (RefPtr protectedThis = weakThis.get())
+                [protectedThis->m_displayLayer enqueueSampleBuffer:sample.get()];
+            ALLOW_DEPRECATED_DECLARATIONS_END
+        });
+    }
+}
+
+void VideoMediaSampleRenderer::flushCompressedSampleQueue()
+{
+    assertIsCurrent(m_workQueue);
+    if (!m_compressedSampleQueue)
+        return;
+
+    PAL::CMBufferQueueReset(m_compressedSampleQueue.get());
+    PAL::CMTimebaseSetTimerDispatchSourceNextFireTime(m_timebase.get(), m_timerSource.get(), PAL::kCMTimeInvalid, 0);
+}
+
+void VideoMediaSampleRenderer::flushDecodedSampleQueue()
+{
+    assertIsCurrent(m_workQueue);
+    if (!m_decodedSampleQueue)
+        return;
+
+    PAL::CMBufferQueueReset(m_decodedSampleQueue.get());
+    PAL::CMTimebaseSetTimerDispatchSourceNextFireTime(m_timebase.get(), m_timerSource.get(), PAL::kCMTimeInvalid, 0);
+}
+
+void VideoMediaSampleRenderer::purgeDecodedSampleQueue()
+{
+    assertIsCurrent(m_workQueue);
+    if (!m_decodedSampleQueue)
+        return;
+
+    if (!m_timebase)
+        return;
+
+    CMTime currentTime = PAL::CMTimebaseGetTime(m_timebase.get());
+    CMTime nextPurgeTime = PAL::kCMTimeInvalid;
+
+    while (auto nextSample = (CMSampleBufferRef)const_cast<void*>(PAL::CMBufferQueueGetHead(m_decodedSampleQueue.get()))) {
+        CMTime presentationTime = PAL::CMSampleBufferGetOutputPresentationTimeStamp(nextSample);
+        CMTime duration = PAL::CMSampleBufferGetOutputDuration(nextSample);
+        CMTime presentationEndTime = PAL::CMTimeAdd(presentationTime, duration);
+        if (PAL::CMTimeCompare(presentationEndTime, currentTime) >= 0) {
+            nextPurgeTime = presentationEndTime;
+            break;
+        }
+
+        RetainPtr sampleToBePurged = adoptCF(PAL::CMBufferQueueDequeueAndRetain(m_decodedSampleQueue.get()));
+        sampleToBePurged = nil;
```

Comment:
unneeded, sampleToBePurged will go out of scope

Reply:
Doing so will result in a "unused variable" warning.